core/store/src/flat/mod.rs

```diff
@@ -0,0 +1,151 @@
+//! FlatStorage is created as an additional representation of the state alongside with Tries.
+//! It simply stores a mapping for all key value pairs stored in our Tries (leaves of the trie)
+//! It is used for fast key value look up in the state. Reading a single key/value in trie
+//! requires traversing the trie from the root, loading many nodes from the database. In flat storage,
+//! we store a mapping from keys to value references so that key value lookup will only require two
+//! db accesses - one to get value reference, one to get the value itself. In fact, in the case of small
+//! values, flat storage only needs one read, because value will be stored in the mapping instead of
+//! value ref.
+//!
+//! The main challenge in the flat storage implementation is that we need to able to handle forks,
+//! so the flat storage API must support key value lookups for different blocks.
+//! To achieve that, we store the key value pairs of the state on a block (head of the flat storage)
+//! on disk and also store the change deltas for some other blocks in memory. With these deltas,
+//! we can perform lookups for the other blocks. See comments in `FlatStorage` to see
+//! which block should be the head of flat storage and which other blocks do flat storage support.
+//!
+//! This file contains the implementation of FlatStorage. It has three essential structs.
+//!
+//! `FlatStorageChunkView`: this provides an interface to get value or value references from flat storage. This
+//!              is the struct that will be stored as part of Trie. All trie reads will be directed
+//!              to the flat state.
+//! `FlatStorageManager`: this is to construct flat state.
+//! `FlatStorage`: this stores some information about the state of the flat storage itself,
+//!                     for example, all block deltas that are stored in flat storage and a representation
+//!                     of the chain formed by these blocks (because we can't access ChainStore
+//!                     inside flat storage).
+
+#[cfg(feature = "protocol_feature_flat_state")]
+pub mod chunk_view;
+pub mod delta;
+#[cfg(feature = "protocol_feature_flat_state")]
+pub mod manager;
+pub mod storage;
+#[cfg(feature = "protocol_feature_flat_state")]
+pub mod store_helper;
+pub mod types;
```

Comment:
Can we make some of these private? I don't know myself, but perhaps for `manager` exposing `pub use manager::FlatStorageManager` is enough.
Ideally `store_helper` must be private, but I'm sure it is not possible without more refactoring.

Reply:
removed `pub` where possible: 7315ab5db
`store_helper` is unfortunately used in chain directly, so right now we can't make it private.